<div style="border:solid Chocolate 2px; padding: 40px">

<b> Александр, привет!👋</b>

Меня зовут Алексей Гриб, я буду твоим ментором на протяжении финального спринта:)
    
Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель нашей работы - не искать ошибки в твоём проекте, а помочь тебе сделать его ещё лучше, устранив недочёты и приблизив его к реальным задачам аналитика. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ментора <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ментора <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
критичные ошибки, которые обязательно нужно исправить.
</div>

Я не смогу принять проект, если в нём будет хотя бы одна критичная ошибка или несколько некритичных ошибок - тогда проект нужно будет немного доработать. Но это нестрашно - я обязательно дам тебе подсказку или укажу верное направление.
    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

# Проект по SQL

## Информация по проекту

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке.

## Цель

Проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание таблиц

Таблица `books`

Содержит данные о книгах:

* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.  

Таблица `authors`

Содержит данные об авторах:

* `author_id` — идентификатор автора;
* `author` — имя автора.  

Таблица `publishers`

Содержит данные об издательствах:

* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства.  

Таблица `ratings`

Содержит данные о пользовательских оценках книг:

* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.  

Таблица `reviews`

Содержит данные о пользовательских обзорах:

* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `usernam`e — имя автора обзора;
* `text` — текст обзора.

## Задания

Выполнять, делая по одному SQL-запросу  
### Посчитайте, сколько книг вышло после 1 января 2000 года;  
### Для каждой книги посчитайте количество обзоров и среднюю оценку;  
### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.  

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Хорошее вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

## Выполнение заданий

### Импортируем библиотеки

In [1]:
import pandas as pd
import sqlalchemy as sa

### Настраиваем подключение

In [2]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали, коннектор сделали - отлично!

### Исследуем таблицы

In [3]:
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for table_name in tables:
    query = f"SELECT * FROM {table_name} LIMIT 5"
    display(get_sql_data(query))

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [4]:
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for table_name in tables:
    query = f"SELECT COUNT(*) FROM {table_name}"
    display(f"В таблице {table_name} строк:")
    display(get_sql_data(query))
    display()

'В таблице books строк:'

count
0   1000

'В таблице authors строк:'

count
0    636

'В таблице publishers строк:'

count
0    340

'В таблице ratings строк:'

count
0   6456

'В таблице reviews строк:'

count
0   2793

**Вывод**  
- в данных представлено 5 таблиц: 
 - в таблице `books` 1000 строк
 - в таблице `authors` 636 строк
 - в таблице `publishers` 340 строк 
 - в таблице `ratings` 6456 строк
 - в таблице `reviews` 2793 строки

### Задание №1

Посчитайте, сколько книг вышло после 1 января 2000 года;

In [5]:
query = '''
        SELECT count(book_id)
        FROM books
        WHERE publication_date > '2000-01-01' 
        '''
get_sql_data(query)

count
0    819

- После 1 января 2000 года вышло 819 книг

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Отмечу, что условие фильтрации `publication_date >= '2000-01-01' ` также было бы правильным, хоть и привело бы к другому результату - в реальных проектах о том, включается ли дата в диапазон или нет, всегда стоит уточнять заранее:)

</div>

### Задание №2

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [6]:
query = '''
WITH reviews_count as (SELECT book_id,
                              count(username) as amount_rev
                       FROM reviews
                       GROUP BY book_id),
                       
     avg_rating as (SELECT book_id,
                           ROUND(AVG(rating),2) as avg_rate
                    FROM ratings
                    GROUP BY book_id)
                    
SELECT b.title,
       rc.amount_rev,
       ar.avg_rate
FROM books as b
LEFT JOIN reviews_count as rc ON b.book_id = rc.book_id
LEFT JOIN avg_rating as ar ON b.book_id = ar.book_id
ORDER BY amount_rev DESC,
         avg_rate DESC
'''
display(get_sql_data(query).head(10))
get_sql_data(query).tail(10)

title  amount_rev  avg_rate
0   The Cat in the Hat and Other Dr. Seuss Favorites         NaN      5.00
1  Disney's Beauty and the Beast (A Little Golden...         NaN      4.00
2                               Leonardo's Notebooks         NaN      4.00
3                          Essential Tales and Poems         NaN      4.00
4    Anne Rice's The Vampire Lestat: A Graphic Novel         NaN      3.67
5                            The Natural Way to Draw         NaN      3.00
6                            Twilight (Twilight  #1)         7.0      3.66
7  Harry Potter and the Prisoner of Azkaban (Harr...         6.0      4.41
8  Harry Potter and the Chamber of Secrets (Harry...         6.0      4.29
9                                     The Book Thief         6.0      4.26

title  amount_rev  avg_rate
990                           The Social Contract         1.0      3.50
991  The Calvin and Hobbes Tenth Anniversary Book         1.0      3.50
992                                 The Bad Place         1.0      3.50
993                             The Bachman Books         1.0      3.33
994                 Debt of Honor (Jack Ryan  #7)         1.0      3.00
995                           Moo  Baa  La La La!         1.0      3.00
996                Death: The High Cost of Living         1.0      3.00
997                             History of Beauty         1.0      2.50
998          Tsubasa: RESERVoir CHRoNiCLE  Vol. 1         1.0      2.50
999            1 000 Places to See Before You Die         1.0      2.50

 - Посчитали для каждой книги количество обзоров и средний рейтинг
 - Имеются книги без обзоров 
 - Максимальное количество обзоров - 7
 - минимальное количетсво обзоров -1

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Метрики для книг посчитаны верно! Здорово, что метрики посчитаны и для тех книг, для которых нет отзывов - довольно частая ошибка в этом задании:)

</div>

### Задание №3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [7]:
query = '''
WITH publish as (SELECT b.book_id,
                        p.publisher
                 FROM books as b
                 JOIN publishers as p ON b.publisher_id = p.publisher_id
                 WHERE b.num_pages > 50)
                 
SELECT publisher,
       COUNT(book_id) as amount_books
FROM publish
GROUP BY publisher
ORDER BY COUNT(book_id) DESC
LIMIT 1
'''
get_sql_data(query)

publisher  amount_books
0  Penguin Books            42

 - Издательство `Penguin Books` выпустило 42 книги толщиной больше 50 страниц

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Издательство опредили верно! Здорово, что дополнительно выведено количество книг издательства:)

</div>

### Задание №4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [8]:
query = '''
WITH book_rate AS (SELECT r.book_id,
                             AVG(r.rating) AS avg_rating,
                             COUNT(r.rating) AS total_ratings
                   FROM ratings r
                   GROUP BY r.book_id
                   HAVING COUNT(r.rating) >= 50),
                      
     author_avg_rate AS (SELECT a.author_id,
                                SUM(br.total_ratings) AS total_ratings,
                                AVG(br.avg_rating) AS avg_rating
                         FROM authors a
                         JOIN books b ON a.author_id = b.author_id
                         JOIN book_rate br ON b.book_id = br.book_id
                         GROUP BY a.author_id)
                                
SELECT a.author,
       aar.total_ratings,
       ROUND(aar.avg_rating,2) AS avg_rate
FROM authors a
JOIN author_avg_rate aar ON a.author_id = aar.author_id
ORDER BY aar.avg_rating DESC
LIMIT 1
'''
get_sql_data(query)

author  total_ratings  avg_rate
0  J.K. Rowling/Mary GrandPré          310.0      4.28

- Автор с самой высокой оценкой книг `J.K. Rowling/Mary GrandPré`
- количество оценок 310
- средний рейтинг 4.28

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Автор с самым высоким средним рейтингом найден верно! Классно, что дополнительно выведен сам средний рейтинг - именно в его расчёте обычно делают ошибки:)

</div>

### Задание №5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [9]:
query = '''
WITH user_rate AS (SELECT username,
                          COUNT(rating_id) AS rate_count
                   FROM ratings
                   GROUP BY username
                   HAVING COUNT(rating_id) > 48),
                                 
user_reviews AS (SELECT username,
                        COUNT(review_id) AS reviews_count
                 FROM reviews
                 WHERE username IN (SELECT username FROM user_rate)
                                   GROUP BY username)
                 
SELECT ROUND(AVG(reviews_count),2) AS avg_reviews
FROM user_reviews
'''
get_sql_data(query)

avg_reviews
0         24.0

 - 24 - Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, посчитано верно!

**Выводы:** 

- в данных представлено 5 таблиц: 
 - в таблице `books` 1000 строк
 - в таблице `authors` 636 строк
 - в таблице `publishers` 340 строк 
 - в таблице `ratings` 6456 строк
 - в таблице `reviews` 2793 строки  
- После 1 января 2000 года вышло 819 книг
- Посчитали для каждой книги количество обзоров и средний рейтинг
 - Имеются книги без обзоров 
 - Максимальное количество обзоров - 7
 - минимальное количетсво обзоров -1
- Издательство `Penguin Books` выпустило 42 книги толщиной больше 50 страниц
- Автор с самой высокой оценкой книг `J.K. Rowling/Mary GrandPré`
 - количество оценок 310
 - средний рейтинг 4.28
- 24 - Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ментора: итоговый вывод по проекту.

Александр, вот мы и преодолели финишную черту - поздравляю с завершением финального спринта!
    
Проделана без сомнения крутая и тяжёлая работа, и все проекты финального спринта удадось выполнить на достойном уровне - уверен, если решишь развиваться дальше в направлении аналитики, тебя ждёт успех и хороший рост:)
    
Спасибо за хорошую работу над проектами финального спринта, желаю дальнейшего профессионального роста и развития, проектов для применения на практике полученных значений, а также адекватных заказчиков:)